src/vanilla/store.ts

```diff
@@ -487,42 +485,66 @@ export const createStore = () => {
     }
-    const loop1 = (a: AnyAtom) => {
-      getDependents(a).forEach((dependent) => {
-        if (dependent !== a) {
-          dependencyMap.set(
-            dependent,
-            (dependencyMap.get(dependent) || new Set()).add(a),
-          )
-          dirtyMap.set(dependent, (dirtyMap.get(dependent) || 0) + 1)
-          loop1(dependent)
-        }
-      })
+
+    // This is a topological sort via depth-first search, slightly modified from
+    // what's described here for performance reasons:
+    // https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+
+    // Step 1: find all atoms in the graph
+    const allAtomsInDependencyGraph = new Array<AnyAtom>()
+    const seen = new Set<AnyAtom>()
+    const queue = [atom]
+    let a: AnyAtom | undefined
+    while ((a = queue.pop())) {
+      if (!seen.has(a)) {
+        seen.add(a)
+        allAtomsInDependencyGraph.push(a)
+        queue.push(...getDependents(a))
+      }
     }
-    loop1(atom)
-    const loop2 = (a: AnyAtom) => {
-      getDependents(a).forEach((dependent) => {
-        if (dependent !== a) {
-          let dirtyCount = dirtyMap.get(dependent)
-          if (dirtyCount) {
-            dirtyMap.set(dependent, --dirtyCount)
-          }
-          if (!dirtyCount) {
-            let isChanged = !!dependencyMap.get(dependent)?.size
-            if (isChanged) {
-              const prevAtomState = getAtomState(dependent)
-              const nextAtomState = readAtomState(dependent, true)
-              isChanged = !isEqualAtomValue(prevAtomState, nextAtomState)
-            }
-            if (!isChanged) {
-              dependencyMap.forEach((s) => s.delete(dependent))
-            }
-          }
-          loop2(dependent)
+
+    // Step 2: traverse the dependency graph to build the topsorted atom list
+    // We don't bother to check for cycles, which simplifies the algorithm.
+    const topsortedAtoms = new Array<AnyAtom>(allAtomsInDependencyGraph.length)
+    let topsortedAtomsIndex = allAtomsInDependencyGraph.length
+    const markedAtoms = new Set<AnyAtom>()
+    const visit = (n: AnyAtom) => {
+      if (markedAtoms.has(n)) {
+        return
+      }
+      markedAtoms.add(n)
+      for (const m of getDependents(n)) {
+        if (n !== m) {
+          visit(m)
         }
-      })
+      }
+      // The algorithm calls for pushing onto the front of the list, so we fill
+      // the array in reverse order.
+      topsortedAtoms[--topsortedAtomsIndex] = n
+    }
+    while (allAtomsInDependencyGraph.length) {
+      visit(allAtomsInDependencyGraph.pop()!)
+    }
+
+    // Step 3: use the topsorted atom list to recompute all affected atoms
+    // Track what's changed, so that we can short circuit when possible
+    const changedAtoms = new Set<AnyAtom>([atom])
+    for (const a of topsortedAtoms) {
+      const prevAtomState = getAtomState(a)
+      let hasChangedDeps = false
+      for (const dep of prevAtomState?.d.keys() || []) {
```

Comment:
On second thought, an early return may look better and performant.
```js
if (!prevAtomState) {
  continue
}
```

Reply:
Yep, done